x-pack/plugins/fleet/server/routes/epm/handlers.ts

```diff
@@ -113,45 +112,50 @@ export const getFileHandler: RequestHandler<TypeOf<typeof GetFileRequestSchema.p
     const useLocalFile = pkgVersion === installation?.version;
 
     if (useLocalFile) {
-      const archiveKey = `${pkgName}-${pkgVersion}/${filePath}`;
-      const archiveEntry = getArchiveEntry(archiveKey);
-      const assetSavedObject = await savedObjectsClient.get<PackageAsset>(
-        ASSETS_SAVED_OBJECT_TYPE,
-        assetPathToObjectId(archiveKey)
-      );
+      const assetPath = `${pkgName}-${pkgVersion}/${filePath}`;
+      const fileBuffer = getArchiveEntry(assetPath);
+      // only pull local installation if we don't have it cached
+      const storedAsset = !fileBuffer && (await getAsset({ savedObjectsClient, path: assetPath }));
 
-      if (!archiveEntry && !assetSavedObject) {
+      // error, if neither is available
+      if (!fileBuffer && !storedAsset) {
         return response.custom({
           body: `installed package file not found: ${filePath}`,
           statusCode: 404,
         });
       }
 
-      const headerContentType =
-        assetSavedObject.attributes.media_type || mime.contentType(path.extname(archiveKey));
-      if (!headerContentType) {
+      // if storedAsset is not available, fileBuffer *must* be
+      // b/c we error if we don't have at least one, and storedAsset is the least likely
+      const { buffer, contentType } = storedAsset
```

Comment:
Feels like there's some relation / overlap with [`ArchiveEntry`](https://github.com/elastic/kibana/blob/4e9afeebd5a4d1406ef23217cdd1821312c4dc40/x-pack%2Fplugins%2Ffleet%2Fserver%2Fservices%2Fepm%2Farchive%2Findex.ts#L25-L28) but it's not yet clear to me.

With `path`, `buffer` and `contentType` we can do a lot.